rpc/common.go

```diff
@@ -7,12 +7,21 @@ import (
 	"bytes"
 	"encoding/gob"
 	"fmt"
+	"log"
 	"net/rpc"
+	"reflect"
 
 	"github.com/hashicorp/hcl/v2/hcldec"
 	"github.com/zclconf/go-cty/cty"
+	"google.golang.org/protobuf/proto"
 )
 
+// UseProto lets us determine whether or not we should use protobuf for serialising
+// data over RPC instead of gob.
+//
+// This is controlled by Packer using the `--use-proto` flag on plugin commands.
+var UseProto bool = false
```

Comment:
I believe we should store this information on the server and client as fields. As opposed to using this global variable. 
In Packer when we check if the plugins support protocol_version "v2" we can config the client using PluginClientConfig to UseProto. The server as I mentioned above when passed the `--protobuf` flag will create the server with the UseProto field set to true.  


The code for the client and server in common then start to look like 

```go
 func (p *commonClient) ConfigSpec() hcldec.ObjectSpec {
	// Legacy: this will need to be removed when we discontinue gob-encoding
	//
	// This is required for backwards compatibility for now, but using
	// gob to encode the spec objects will fail against the upstream cty
	// library, since they removed support for it.
	//
	// This will be a breaking change, as older plugins won't be able to
	// communicate with Packer any longer.
	if !c.UseProto {  // or if c.ProtocolVersion != "v2"
		log.Printf("[DEBUG] - common: receiving ConfigSpec as gob")
		res := hcldec.ObjectSpec{}
		err := gob.NewDecoder(bytes.NewReader(resp.ConfigSpec)).Decode(&res)
		if err != nil {
			panic(fmt.Errorf("failed to decode HCL spec from gob: %s", err))
		}
		return res
	}
}

....

func (s *commonServer) ConfigSpec(_ interface{}, reply *ConfigSpecResponse) error {
// Legacy: this will need to be removed when we discontinue gob-encoding
	//
	// This is required for backwards compatibility for now, but using
	// gob to encode the spec objects will fail against the upstream cty
	// library, since they removed support for it.
	//
	// This will be a breaking change, as older plugins won't be able to
	// communicate with Packer any longer.
	if !s.UseProto {  // or if s.ProtocolVersion != "v2"
		log.Printf("[DEBUG] - common: receiving ConfigSpec as gob")
		res := hcldec.ObjectSpec{}
		err := gob.NewDecoder(bytes.NewReader(resp.ConfigSpec)).Decode(&res)
		if err != nil {
			panic(fmt.Errorf("failed to decode HCL spec from gob: %s", err))
		}
		return res
	}
}

```

I have to dive deeper into the datasources but I believe it embeds a Common client and a  Common server if so I believe the fields may be accessible on the datasource.

Reply:
> I'm not sure as to the proposed change though, both clients and servers need to be in line with the protocol used, the current approach means UseProto can be global, it however needs to be set to either true or false ASAP in the lifecycle of the processes, that means at command start for plugins, and once the discovery is done for Packer.

This is correct. The caller would be responsible for initializing the server and client with the protocol enabled so since Packer is the control here. When we identify that all plugins support protocol v2 we setup Packer to start create a client for the plugin with the protocol set to use protobuf. The execute command that starts the plugin with the --protobuf flag. This will tell the SDK that the server should use protobuf. At this point the client and server will be in line. 


> E.g. from a plugin perspective, I'm pretty certain we won't have access to the command-flags at the point where we create the RPC structs, so we'll probably need to either forward it, or keep it globally. If we keep it global, are there merits to this approach, as we'll copy a global flag into a structure?